sessionctx/variable/variable.go

```diff
@@ -188,6 +189,11 @@ func (sv *SysVar) GetSessionFromHook(s *SessionVars) (string, error) {
 
 // SetSessionFromHook calls the SetSession func if it exists.
 func (sv *SysVar) SetSessionFromHook(s *SessionVars, val string) error {
+	if sv.Name == MaxAllowedPacket {
```

Comment:
Please move the code snippet as the `SetSession()` function of [`MaxAllowedPacket`](https://github.com/pingcap/tidb/blob/69f664ae051c21e0edc48fcc5556f19fe4465a9a/sessionctx/variable/sysvar.go#L845)

Reply:
If the code snippet is moved to `SetSession()`, the truncation of `val` would be invalid in `SetSessionFromHook`, so the statement [`s.systems[sv.Name] = val`](https://github.com/pingcap/tidb/blob/69f664ae051c21e0edc48fcc5556f19fe4465a9a/sessionctx/variable/variable.go#L202) would set back the un-truncated value.

Reply:
I see.

It is a concern for me that we have an explicit `if` for `MaxAllowedPacket`, I think those codes should be `SysVar`-specific rather than defined in `SetSessionFromHook`.

@morgo Do you have any suggestion?

Reply:
Yeah, I would like to move it to the sysvar as well.

Sorry, I'm not sure I understand the problem yet. Can you do the truncation in the `Validation()` func? (it should be able to change values if required in validation)